Makefile

```diff
@@ -272,20 +273,31 @@ get-ci-tools:
 
 # Run go fmt against code
 gofmt: get-ci-tools
-	$(CI_TOOLS_REPO_DIR)/test-runner/gofmt.sh
+	GOWORK=off $(CI_TOOLS_REPO_DIR)/test-runner/gofmt.sh
+	GOWORK=off $(CI_TOOLS_REPO_DIR)/test-runner/gofmt.sh ./api
 
 # Run go vet against code
 govet: get-ci-tools
-	$(CI_TOOLS_REPO_DIR)/test-runner/govet.sh
+	GOWORK=off $(CI_TOOLS_REPO_DIR)/test-runner/govet.sh
+	GOWORK=off $(CI_TOOLS_REPO_DIR)/test-runner/govet.sh ./api
 
 # Run go test against code
 gotest: get-ci-tools
-	$(CI_TOOLS_REPO_DIR)/test-runner/gotest.sh
+	GOWORK=off $(CI_TOOLS_REPO_DIR)/test-runner/gotest.sh
+	GOWORK=off $(CI_TOOLS_REPO_DIR)/test-runner/gotest.sh ./api
 
 # Run golangci-lint test against code
 golangci: get-ci-tools
-	$(CI_TOOLS_REPO_DIR)/test-runner/golangci.sh
+	GOWORK=off $(CI_TOOLS_REPO_DIR)/test-runner/golangci.sh
+	GOWORK=off $(CI_TOOLS_REPO_DIR)/test-runner/golangci.sh ./api
 
 # Run go lint against code
 golint: get-ci-tools
-	PATH=$(GOBIN):$(PATH); $(CI_TOOLS_REPO_DIR)/test-runner/golint.sh
+	GOWORK=off PATH=$(GOBIN):$(PATH); $(CI_TOOLS_REPO_DIR)/test-runner/golint.sh
+	GOWORK=off PATH=$(GOBIN):$(PATH); $(CI_TOOLS_REPO_DIR)/test-runner/golint.sh ./api
+
+.PHONY: gowork
+gowork: ## Generate go.work file to support our multi module repository
+	test -f go.work || go work init
+	go work use .
+	go work use ./api
```

Comment:
-1: I would add the go.work directly in the repo instead of having this line

Reply:
So you're suggesting running those 3 commands and put the resulting files (`go.work` and `go.work.sum`) within the repository: am I understanding correctly?
By doing that we can avoid this target here.

Reply:
FYI, we added this target to other operator Makefile as the design workspace doc says https://go.googlesource.com/proposal/+/master/design/45713-workspace.md#rationale-the-file

~~~
These go.work files should not be checked into the repositories so that they don‘t override the workspaces users explicitly define. Checking in go.work files could also lead to CI/CD systems not testing the actual set of version requirements on a module and that version requirements among the repository’s modules are properly incremented to use changes in the modules.
~~~

Why we could do different, this was the initial reason why we did not.

Reply:
> Instead of `go mod edit -replace` like we suggest I think it's better to do `go work edit -replace` because that way our local changes will be ignored by git since go.work is in .gitignore That's what I'm going to suggest in the docs.

When a commit depends on another commit in another repository I tend to actually check in and publish the `replace` in go.mod so that when people review the PR they can pull it and run it locally with the correct deps. Sure this means once the dependency merges I need to respin my PR but overall but this is my low tech version of zuul's `Depends-On:`

Reply:
+1 and thanks @gibizer for sharing. The whole point of having `go.work` (in my recent experience) is to make the `go mod edit -replace` working properly when a local `lib-common` is referenced in the project.
I didn't find a nice way to make it working without creating a go workspace.
It would be great (in the next future) having our CI able to checkout the dependencies and run a job which takes into account the PR dependencies.

Reply:
@stuggi Iiuc that paragraph is referring to not checking in the `go.work` with actual replacements, which is also what I'm suggesting.

By having the `go.work` with only the `.` and `./api` in the repo and excluding it with `.gitignore` we avoid anyone from inadvertently submitting it.

@gibizer I myself don't like the PRs having the `go.mod` changes included.  I have my own python script where I can tell it the PR and it does the replacing automatically. For example for the extra volume stuff where we use the lib-common PR#88 and openstack-operator PR#38 I just run:

```
hack/replace.py lib-common=88 openstack-operator=38
```

And that makes the right replacements in `go.work` for me.

It should be pretty straightforward building the `Depend-On` with that, because the script also accepts local directories and even specific branches of repos, where repos can be long or short version.  For example the short version for the previous one would be:

```
hack/replace.py lib-common=fmount/lib-common@extra_volumes openstack-operator=fmount/openstack-operator@extra_volumes
```

Reply:
> @gibizer I myself don't like the PRs having the `go.mod` changes included. I have my own python script where I can tell it the PR and it does the replacing automatically. For example for the extra volume stuff where we use the lib-common PR#88 and openstack-operator PR#38 I just run:
> 
> ```
> hack/replace.py lib-common=88 openstack-operator=38
> ```
> 
> And that makes the right replacements in `go.work` for me.
> 
> It should be pretty straightforward building the `Depend-On` with that, because the script also accepts local directories and even specific branches of repos, where repos can be long or short version. For example the short version for the previous one would be:
> 
> ```
> hack/replace.py lib-common=fmount/lib-common@extra_volumes openstack-operator=fmount/openstack-operator@extra_volumes
> ```

I guess it is matter of taste. I like my way as it makes explicit for the reader of the PR that there is a dependency to a specified version and location (i.e.xyz branch of the fork).
By all means I'm supportive to automate dependency handling via some scripting.

Reply:
I like to have it explicitly stated in the commit message (like with th e`Depends-on` you mention).

Reply:
> @stuggi Iiuc that paragraph is referring to not checking in the `go.work` with actual replacements, which is also what I'm suggesting.
> 
> By having the `go.work` with only the `.` and `./api` in the repo and excluding it with `.gitignore` we avoid anyone from inadvertently submitting it.
> 

right, we definitely want to have it in `.gitignore` to prevent people to add it to commits. Having a default checked in is something we could do, but its not common practice. So far I have not found other projects which do this. In our case for the main and api sub module its probably ok to add the go.work to the repo as they depend on each other.

Reply:
Ok, I see @stuggi's point of view and in general I agree about not having `go.work` and `go.work.sum` in the repos.
However, I understand that PR #77  mentioned by @Akrog introduces a series of scripts with the goal of simplify the developer life, automating the way we checkout PRs [1] and set dependencies to `GOWORK` [2] in cinder-operator.
In this case I suspect it's good to maintain consistency with the direction taken by this operator, and we can actually commit a `go.work` file (which is still in `.gitignore`), removing the need to execute the `gowork` target each time.
We might want to keep the `gowork` target in the `Makefile` (it can be explicitly called if needed), but I can remove its execution in the `vet` target (and wherever it occurs).

I'm looking for quorum to move forward with this, so I'd like to have your feedback @stuggi @Akrog 


[1] https://github.com/openstack-k8s-operators/cinder-operator/pull/77/files#diff-adc3e455779ae5d167129a8b2a045ac46868c35ea47fcfb127327f0c3ba7cf45
[2] https://github.com/openstack-k8s-operators/cinder-operator/pull/77/files#diff-58fb17599cdc1f9bd44fbdb460b74f3de822da6a4f5007f0ac97831ee0e4ec68

Reply:
I am fine with the proposal, adding the default `go.work`  with having it in `.gitignore` and remove calling it in the targets.

Reply:
If we really don't want to have the go.work file included to be the same as other projects I could modify my scripts to create it when necessary, though we would still want to default `GOWORK` to off